R/tool-memory.R

```diff
@@ -0,0 +1,428 @@
+        items = ellmer::type_string()
+      )
+    )
+  }
+)
+
+#' Read btw memory YAML file
+#'
+#' Reads the btw-memory.yaml file and returns a btw_memory object.
+#' If the file doesn't exist, returns an empty btw_memory structure.
+#'
+#' @param path Path to the YAML file. Defaults to "btw-memory.yaml"
+#'
+#' @return A list with class "btw_memory" containing `project_context` and
+#'   `data_sources`
+#'
+#' @noRd
+read_btw_memory_yaml <- function(path = NULL) {
+  path <- path_find_btw_memory(path)
+
+  as_btw_memory(yaml::read_yaml(path) %||% list())
+}
+
+as_btw_memory <- function(x) {
+  structure(x, class = "btw_memory")
+}
+
+#' Write btw memory YAML file
+#'
+#' Writes a btw_memory object to a YAML file.
+#'
+#' @param x A btw_memory object (list with project_context and data_sources)
+#' @param path Path to the YAML file. Defaults to "btw-memory.yaml"
+#' @noRd
+write_btw_memory_yaml <- function(x, path = NULL) {
+  if (!inherits(x, "btw_memory")) {
+    stop("Object must have class 'btw_memory'")
+  }
+
+  data <- unclass(x)
+  path <- path_find_btw_memory(path, must_exist = FALSE)
+
+  if (identical(compact(data), list())) {
+    cli::cli_inform(
+      "No memory to write, creating an empty file at {.path {path}}."
+    )
+    fs::file_touch(path)
+    return(invisible(x))
+  }
+
+  # Project Context ----
+  if (!is.null(data$project_context)) {
+    # problem_description should be a single string
+    if (!is.null(data$project_context$problem_description)) {
+      data$project_context$problem_description <- paste(
+        data$project_context$problem_description,
+        collapse = " "
+      )
+    }
+  }
+
+  # Data Sources ----
+  if (!is.null(data$data_sources) && length(data$data_sources) > 0) {
+    for (i in seq_along(data$data_sources)) {
+      ds <- data$data_sources[[i]]
+
+      # These should be single strings
+      for (field in c("name", "description", "source", "code")) {
+        if (!is.null(ds[[field]])) {
+          ds[[field]] <- paste(ds[[field]], collapse = "\n")
+        }
+      }
+
+      # notes should be an array of strings
+      if (!is.null(ds$notes)) {
+        ds$notes <- as.character(ds$notes)
+      }
+
+      # Variables ----
+      if (!is.null(ds$variables) && length(ds$variables) > 0) {
+        for (j in seq_along(ds$variables)) {
+          var <- ds$variables[[j]]
+
+          # name should be a single string
+          if (!is.null(var$name)) {
+            var$name <- paste(var$name, collapse = "")
+          }
+
+          # notes should be an array of strings
+          if (!is.null(var$notes)) {
+            var$notes <- as.character(var$notes)
+          }
+
+          ds$variables[[j]] <- var
+        }
+      }
+
+      data$data_sources[[i]] <- ds
+    }
+  }
+
+  yaml::write_yaml(data, path, indent.mapping.sequence = TRUE, indent = 2)
+
+  invisible(x)
+}
+
+
+path_find_btw_memory <- function(path = NULL, must_exist = TRUE) {
```

Comment:
Noting that in the `list_files` tool, we've just considered the working directory as the project directory:

https://github.com/posit-dev/btw/blob/ee33182ce2bbfbc658ab5a28e7752e8b9da19d9e/tests/testthat/_snaps/tool-files.md?plain=1#L22-L25